### PR TITLE
Refactor with vanilla

### LINKED ISSUE #220: UseEffect not invoked on atom value change in special circumstances
Follow-up to https://github.com/pmndrs/jotai/issues/208 (via https://github.com/pmndrs/jotai/pull/215).

UseEffect sometimes doesn't invoke with updated atom value, despite it being listed in dependencies. There's some weird race condition that probably cancels it or something. Don't understand what's happening and repro itself is insane.

Repro: https://codesandbox.io/s/morning-night-fhwvh?file=/src/index.js&expanddevtools=1

- **Jotai version:** https://pkg.csb.dev/pmndrs/jotai/commit/b989cdbf/jotai (the one after `0.11.1`)
- **React versions:**  `17.0.0`
- **Reproduces in strict mode:** yes (both strict and non-strict)

### LINKED ISSUE #235: "Primitive" vs. "primitive" atoms
I'm not sure if this is a bug. Even if it is, I'm not sure it's worth fixing---it may be the best you can manage with typescript. But it did confuse me for a bit. 

The documentation talks about "primitive" atoms, which are atoms created with an initial value:
```
const anAtom = atom('foo')
```
There is also a public type `PrimitiveAtom<T>`. All "primitive" atoms are of type `Primitive<T>`, but you can create a derived atom and give it type PrimitiveAtom<T> without it actually being  "primitive". E.g.,
```
const pAtom: PrimitiveAtom<string> = atom(
  (get) => get(someOtherAtom)
  (get, set, update: SetStateAction<string>) => {set(someOtherAtom, update)}
)
```
I got confused because I knew I needed a "primitive" atom for some purpose, and thought that any object with type `PrimitiveAtom` would work. But that wasn't the case. Everything type-checked okay, but it didn't actually work at run time.

### PR DESCRIPTION
**Pull request by dai-shi**
This is a complete refactoring of the core.

The primary goal is to make the core code more readable and maintainable.
Split the core module out of Provider.ts into vanilla.ts without React dependency.
Now, the code is much cleaner with single WeakMap, and unified state.
All tests are passing with some tiny exceptions.

There are some changes.
- We avoid Promise.resolve() batching and let React batch updates. (Fix a bug with `<input>` which has been existed since v0.10.0)
- Drop `atom.key` and add `atom.toString()` instead. (To be documented.)
- Fix #220 
- Better `PrimitiveAtom` typing, close #235

### CODE REVIEW

**Comment by dai-shi** [comment]
#220 
https://codesandbox.io/s/eloquent-kirch-zwy7v?file=/src/index.js
Should be fixed.
cc @dwelle

**Comment by dwelle** [comment]
Sadly I can't repro the original bug anymore (I've saved the original commit) :(. But the CSB repro is fixed, so that's great :tada:  :heart:

**Comment by dai-shi** [comment]
https://codesandbox.io/s/jotai-r3f-fri9d
jotai with r3f works fine with useBridge/Bridge.

**Comment by dai-shi** [comment]
https://codesandbox.io/s/jotai-todos-ijyxm
jotai todos with `key={String(anAtom)}` works fine.

**Comment by dai-shi** [comment]
This is going to be v0.12.0.

File: tests/error.test.tsx [hunk 1]
```diff
@@ -393,7 +373,6 @@ it('throws an error while updating in effect', async () => {
     </Provider>
   )
 
-  expect(console.error).toHaveBeenCalledTimes(0)
```
Comment (dai-shi) [inline tests/error.test.tsx:396]: This is removed, maybe because it's now not async.

**Pull request merged**